bokehjs/src/coffee/renderer/tile/mercator_tile_source.coffee

```diff
@@ -0,0 +1,241 @@
+      return current if (Math.abs(current - resolution) < Math.abs(previous - resolution))
+      return previous
+    return @get('resolutions').indexOf(closest)
+
+  snap_to_zoom: (extent, height, width, level) ->
+    desired_res = @get('resolutions')[level]
+    desired_x_delta = width * desired_res
+    desired_y_delta = height * desired_res
+
+    [xmin, ymin, xmax, ymax] = extent
+    x_adjust = (desired_x_delta - (xmax - xmin)) / 2
+    y_adjust = (desired_y_delta - (ymax - ymin)) / 2
+
+    return [xmin - x_adjust, ymin - y_adjust, xmax + x_adjust, ymax + y_adjust]
+
+  tms_to_wmts: (x, y, z) ->
+    '''Note this works both ways'''
+    return [x, 2 ** z - 1 - y, z]
+
+  wmts_to_tms: (x, y, z) ->
+    '''Note this works both ways'''
+    return [x, 2 ** z - 1 - y, z]
+
+  pixels_to_meters: (px, py, level) ->
+    res = @get_resolution(level)
+    mx = px * res - @get('x_origin_offset')
+    my = py * res - @get('y_origin_offset')
+    return [mx, my]
+
+  meters_to_pixels: (mx, my, level) ->
+    res = @get_resolution(level)
+    px = (mx + @get('x_origin_offset')) / res
+    py = (my + @get('y_origin_offset')) / res
+    return [px, py]
+
+  pixels_to_tile: (px, py) ->
+    tx = Math.max(Math.ceil(px / parseFloat(@get('tile_size'))) - 1, 0)
+    ty = Math.max(Math.ceil(py / parseFloat(@get('tile_size'))) - 1, 0)
+    return [tx, ty]
+
+  pixels_to_raster: (px, py, level) ->
+    mapSize = @get('tile_size') << level
+    return [px, mapSize - py]
+
+  meters_to_tile: (mx, my, level) ->
+    [px, py] = @meters_to_pixels(mx, my, level)
+    return @pixels_to_tile(px, py)
+
+  get_tile_meter_bounds: (tx, ty, level) ->
+    # expects tms styles coordinates (bottom-left origin)
+    [xmin, ymin] = @pixels_to_meters(tx * @get('tile_size'), ty * @get('tile_size'), level)
+    [xmax, ymax] = @pixels_to_meters((tx + 1) * @get('tile_size'), (ty + 1) * @get('tile_size'), level)
+
+    if xmin? and ymin? and xmax? and ymax?
+      return [xmin, ymin, xmax, ymax]
+    else
+      return undefined
+
+  get_tile_geographic_bounds: (tx, ty, level) ->
+    bounds = @get_tile_meter_bounds(tx, ty, level)
+    [minLon, minLat, maxLon, maxLat] = @utils.meters_extent_to_geographic(bounds)
+    return [minLon, minLat, maxLon, maxLat]
+
+  get_tiles_by_extent: (extent, level, tile_border=1) ->
+    # unpack extent and convert to tile coordinates
+    [xmin, ymin, xmax, ymax] = extent
+    [txmin, tymin] = @meters_to_tile(xmin, ymin, level)
+    [txmax, tymax] = @meters_to_tile(xmax, ymax, level)
+
+    # add tiles which border
+    txmin -= tile_border
+    tymin -= tile_border
+    txmax += tile_border
+    tymax += tile_border
+
+    tiles = []
+    for ty in [tymax..tymin] by -1
+      for tx in [txmin..txmax] by 1
+        tiles.push([tx, ty, level, @get_tile_meter_bounds(tx, ty, level)])
+        tiles = @sort_tiles_from_center(tiles, [txmin, tymin, txmax, tymax])
+    return tiles
+
+  quadkey_to_tile_xyz: (quadKey) ->
```

Comment:
`Quadkey`: a string id for a tile whose length is the same as the tile's zoom level and which is composed of parent tile quadkeys. The quadkey makes it simple to traverse the heirarchy of tiles to find parent and child tiles.

The `QUADKEYTileSource` uses a templated url which includes a `{Q}` variable as opposed to `{Z}/{X}/{Y}`.  Bing / Virtual Earth uses this style of tile url. 

So while XYZ tile sources don't use a quadkey in the url, the quadkey id is helpful for other operations within the mercator (square) tile space.  Specifically, I use quadkeys to quickly find parent tiles.